### DISCUSSION TITLE
gabyhelp feedback

### DISCUSSION THREAD

**Discussion by rsc** [golang/go#67901] [Discussions]
I am experimenting with running a new bot @gabyhelp (which I am writing) in the Go issue tracker to try to help automate various mundane things that a machine can do reasonably well. Some aspects of the work may involve AI/LLMs; some will not. 

Please thumbs up or thumbs down individual comments from the bot depending on whether they are correct/helpful or not. If the bot is getting in the way, we obviously want to know and fix it.

When the bot posts a comment, it will link back to this discussion page, which is a good place to leave non-boolean feedback. Thanks!

This discussion is also a good place to post ideas for things you wish a bot would do for you in the issue tracker.

To snoop around in all the issues where the bot has commented, you can use [this search](https://github.com/golang/go/issues?q=is%3Aissue+is%3Aopen+commenter%3Agabyhelp).

**Comment by randall77** [comment]
#68220 is pretty clearly a dup of #68170, but gabyhelp didn't find it.

**Reply by rsc** [reply]
Thanks! I tried manually expanding the playground link in #68170 into an inline code snippet, and after that #68220 is a close enough match to be considered "related". I have been thinking about doing that anyway to make issues a bit more self-contained. Probably we can do that in the comment fixer before the indexing runs.

**Comment by peterGo** [comment]


**Reply by willfaught** [reply]
>The name gaby is short for “Go AI Bot”

Where does the "y" come from?

**Reply by rsc** [reply]
[gab](https://en.wikipedia.org/wiki/Gab_(social_network)) is taken so i added a y.

**Comment by jfesler** [comment]
i'm glad we focus on the important things.

**Comment by mauri870** [comment]
Gaby posted a response with broken markdown in https://github.com/golang/go/issues/68250#issuecomment-2198261130

**Reply by ainar-g** [reply]
@rsc, another markup failure here: https://github.com/golang/go/issues/68501#issuecomment-2234004830.

**Reply by magical** [reply]
Another markdown failure in https://github.com/golang/go/issues/68716#issuecomment-2265879346. The issue links are fine but all the documentation links are garbled.

**Comment by mvdan** [comment]
The bot got it entirely right in https://github.com/golang/go/issues/68253, although the previous issue wasn't the first suggestion.

**Comment by prattmic** [comment]
In #68282, the bot’s post enabled the reporter to find the solution to their problem and self close their own issue. 🎉

**Comment by rsc** [comment]
The gaby bot is still running on my basement server and the power blipped over the weekend, so it's been off for a few days. (We are working on moving it to a cloud server but not there yet.) Expect a burp of related issue posts as it catches up. Apologies.

**Comment by dsnet** [comment]
https://github.com/golang/go/issues/68472#issuecomment-2231795448 seems to have some malformed formatting related to referencing pkgsite.

**Reply by AndrewHarrisSPU** [reply]
Poking around out of curiosity because this is such a cool project...

The pilcrow `¶` seems like a clue (does it come from go.dev?). Also curious [here](https://github.com/rsc/gaby/blob/a38b6c04575c65427b30af5d8e9362b916193d9e/internal/related/related.go#L228): the title text may be the result ID, or the title of an issue retrieved by looking up the result ID. In these cases where is the title text coming from?

**Reply by ghost** [reply]
Similar problem: https://github.com/golang/go/issues/68501#issuecomment-2234004830

**Reply by Jorropo** [reply]
More https://github.com/golang/go/issues/68482#issuecomment-2232655225

**Reply by ghost** [reply]
To be honest, I think gaby should not pass links to `pkg.go.dev` or at least do it in a separate `Documentation`/`Reference` section.

**Comment by zephyrtronium** [comment]
Would it be feasible to provide the related issue analysis as a separate tool? Then people could use it to check for duplicates before opening an issue at all, and it might be faster than using goissues.org and careful GitHub searches to find issues in conversation.

**Comment by Jorropo** [comment]
I prefer how the bot used to link issues, for example: https://github.com/golang/go/issues/67854#issuecomment-2155785397
Versus now: https://github.com/golang/go/issues/68499#issuecomment-2233903645

The markdown different is:
```
- https://github.com/golang/go/issues/68293
```
versus:
```
- [encoding/xml: accepts various ill-formed XML](https://github.com/golang/go/issues/68293)
```

The first one is parsed by the github UI, it replaces the link text with the issue name (just like the bot manually does) and a helpful SVG based on the issue status (***which the bot does not do***) :
![image](https://github.com/user-attachments/assets/456f86c4-7fa7-4dda-a8e5-e78d48446c50)
![image](https://github.com/user-attachments/assets/ff211e27-e91b-491a-a281-3840c05b275f)
![image](https://github.com/user-attachments/assets/1f12b453-88ce-44de-9fbe-401555814f4b)

The github UI version is also updated automagically in case the issue status change or it is renamed.

**Reply by dmitshur** [reply]
It used to do that, but that renders less nicely in email notifications that GitHub mails. It was changed to the current format to make the email notifications more readable, but it has the downside you mention for viewing the comments on github.com itself.

I would also prefer the previous format since I don’t use GitHub’s email notifications, but each one has trade-offs. (Ideally GitHub should change the email notifications to render more similarly to the web version.)

**Reply by Jorropo** [reply]
Rahhhh tradeoffs :'(
Would you happen to know of an existing bug report to github for the better kind to notifications to render "properly" in emails ?

Also I wouldn't mind if the bot used to emojies to emulate the same behavior but I guess the code would be more involved as it would need to update existing comments once the status of an issue change.

**Reply by fvbommel** [reply]
Does GitHub send email notifications on comment edits?
If it does not, the bot could initially post its comments with the "nice for email" markup, then shortly after edit them with the bare links that look better on the site itself.

**Comment by zpavlinovic** [comment]
`x/vuln`: In #68152, @gabyhelp suggestions were not helpful. This is an issue with a brand new feature so there won't be many, if any, similar issues.

`x/vuln`: In #68100, a first few suggestions are ok. The next several suggestions are not relevant and belong to a different repo `x/net`. Would it make sense to restrict suggestions to the same repo, at least for specialized repos such as `x/vuln`?

**Comment by jfrech** [comment]
I think you should hard-code your bot to ignore "x/pkgsite" issues which are package removal requests.

**Comment by jfrech** [comment]
Is `label:gabywins` collecting successful work employing this tool? The label doesn't have a description.

**Comment by gaby** [comment]
@gabyhelp Hello 👋🏻

**Comment by rolandshoemaker** [comment]
It would be perhaps useful to filter out related issues/documentation if they are explicitly mentioned in the parent comment. In #69025 I reference #19413 as a previous/similar issue, which makes @gabyhelp referencing it as related/similar in the follow-up comment somewhat redundant.

**Comment by AlekSi** [comment]
@gabyhelp thinks that all CVEs are similar: https://github.com/golang/go/issues/69138#issuecomment-2318201174

**Comment by andig** [comment]
@rsc is gaby specific to Go or could it be used for other languages/repos? I haven't been able to find much on Github Marketplace that seems to provide similar functionality.

**Reply by dmitshur** [reply]
@andig You may find https://github.com/golang/oscar#prototype and the rest of https://github.com/golang/oscar#readme relevant.

**Comment by FiloSottile** [comment]
It'd be nice if @gabyhelp skipped issues already mentioned in the body of the issue (and omitted commenting if those were all the results).

**Reply by cespare** [reply]
(Also mentioned [here](https://github.com/golang/go/discussions/67901#discussioncomment-9755616) and [here](https://github.com/golang/go/discussions/67901#discussioncomment-10422309).)

**Comment by mateusz834** [comment]
Why does @gabyhelp link to https://go.dev/pkg/go/ast/?m=old#SortImports instead of https://pkg.go.dev/go/ast#SortImports ?

See https://github.com/golang/go/issues/69694#issuecomment-2380638267

**Reply by iand** [reply]
Also @gabyhelp linked to https://github.com/golang/go/issues/36140 which was closed in favour of https://github.com/golang/go/issues/20818 which is still open. Probably 20818 should also be linked.

**Comment by nsrip-dd** [comment]
@gabyhelp left two identical comments on this issue: https://github.com/golang/go/issues/69747#issuecomment-2389195560 and https://github.com/golang/go/issues/69747#issuecomment-2389583989

**Comment by jba** [comment]
@nsrip-dd That's the result of a change in the way we check for done-ness. The new way should have incorporated the old check, but didn't. We'll fix that.

**Comment by mvdan** [comment]
Can @gabyhelp learn to close empty issues? This should also include issues which copied the template but did not fill it out. For example: https://github.com/golang/go/issues/69795

**Reply by jba** [reply]
Filed https://github.com/golang/oscar/issues/37.

**Comment by Jorropo** [comment]
Some kind of issues are not helpful to give recommendations https://github.com/golang/go/issues/70376#issuecomment-2479535693 altho it's mostly harmless
could be very smart like by predicting click through rate and not posting if it would be too low or as simple as hardcoding `access:`

**Reply by prattmic** [reply]
While I agree that linking to other approver issues isn't helpful, it actually is helpful that it linked to your prior may-start-trybots issue. (Though just having the issue title isn't great since it isn't obvious that it is your issue).

**Reply by Jorropo** [reply]
Fair enough

> Though just having the issue title isn't great since it isn't obvious that it is your issue

Yes got me

**Comment by millerresearch** [comment]
@gabyhelp not helpful on issue #70484 reporting on a bug which was introduced by a specific CL. Maybe if the title includes "after CL ..." or "since CL ..." it's not worth looking for other issues?

**Reply by hyangah** [reply]
Thanks! I see gabyhelp's comment doesn't bring new info in this specific case.

On the other hand, when there are already existing reports or the breakage was actually a reoccurring bug, or if the mentioned CL turned out to be causing other issues, the related issue feature could surface them and be helpful (of course, if that can be done correctly).

**Comment by ALTree** [comment]
In https://github.com/golang/go/issues/70571 (a spec-related issue), in the doc section, the bot linked both the current and the historical go1.17 spec at https://go.dev/doc/go1.17_spec. The 1.17 spec page is only there for historical reasons and in general I don't think there's any benefit in linking it in a discussion about the current version of the language; and it could create confusion.

**Reply by hyangah** [reply]
Thanks! Filed https://github.com/golang/oscar/issues/63

**Comment by seankhliao** [comment]
In https://github.com/golang/go/issues/70569#issuecomment-2499863712 it linked to the same issue it was commenting in.

**Reply by hyangah** [reply]
Thanks! Filed https://github.com/golang/oscar/issues/62

**Comment by Jorropo** [comment]
Example of messages that would help if it added codeblocks & formated:
- https://github.com/golang/go/issues/70842
- https://github.com/golang/go/issues/70987

**Comment by jba** [comment]
You may have noticed that @gabyhelp  is now labeling issues. Feedback welcome.

**Reply by seankhliao** [reply]
How do Bug / BugReport differ?

**Reply by jba** [reply]
They are the same. We changed from Bug to BugReport because we felt Bug was confusing: the purported bug may not actually be one. You shouldn't see Bug on any new issues. I will go through replacing Bug with BugReport, and then delete Bug.

**Reply by jba** [reply]
Bug is gone.

**Reply by seankhliao** [reply]
#71362 i think should be a bug report (like some others when people are too lazy to fill out the bug report template), and I would have expected the bot to make the same judgement (rather than just copying the proposal label).

**Reply by seankhliao** [reply]
I feel like the labels it manages can be somewhat redundant with all the other labels we have

perhaps it could instead manage issue types rather than labels?
https://docs.github.com/en/issues/tracking-your-work-with-issues/configuring-issues/managing-issue-types-in-an-organization

**Reply by apparentlymart** [reply]
I've noticed that on some issues it has added the "LibraryProposal" label without also adding the "Proposal" label.

I'm not sure if that was intentional, but I'd personally like there to be one label that is present on all proposals regardless of what they affect, even if there's _also_ more specific labels like "LibraryProposal", just because that makes it easier to quickly construct a filter to help answer the question "was there already a proposal for something like this?" while excluding all of the bug reports, etc.

**Reply by dolmen** [reply]
@apparentlymart A bug may be related to a proposal without being a proposal. For example #74757 is a bug related to the #ToolProposal feature, but isn't a proposal.

**Comment by AndrewHarrisSPU** [comment]
The [gabybot summary of 71203](https://github.com/golang/go/issues/71203#issuecomment-2593973681) was pretty interesting. The summary seems useful (not impressive, not sharp, not insightful, but plausibly useful). It captured topics well, although I thought it missed some interesting or convincing posts and seemed a bit long. Of course it's very, very cool that this exists and wouldn't be written otherwise.

There are further issues specifically with the "issue comment" citations that seemed worth detailing: Out of 27 (19 unique, I think) "issue comment" citations made by gabybot:

1. Sometimes there is a poor conceptual relationship between the point gabyhelp is on, and the content in mentioned posts:
- discussion themes sub 1: kortshack mention
- discussion themes sub 1: dominikh mention (also, link is marked off-topic)
- discussion themes sub 2: thiht mention

(These are only the cases I thought were objectively clear. There are also fuzzier cases where a conceptual relationship was IMHO too thin, but not absent.)

2. Sometimes gabybot has made reference errors, mentioning the correct participant but linking to the wrong post/participant:
- discussion Themes sub 2: misdirect: doggedOwl mention links to ianlancetaylor comment ([doggedOwl comment](https://github.com/golang/go/issues/71203#issuecomment-2587310441) appears to be distant)
- next steps: earthboundkid mention links to mzzsfy comment (mzzsfy comment is immediately prior to earthboundkid comment)
- next steps: puellanivis mention links to jba comment (puellanivis comment is immediately prior to jba comment)

3. Errata:
- spams mentions of the original post early on (7 times)
- discussion themes sub 3: ConradIrwin post linked twice
- next steps: offtopic: godcong mention links to off-topic comment (ironically, comment suggests AI summary)

**Reply by apparentlymart** [reply]
I was kinda amused when this summary appeared while I had already been working on [a manual summary in an earlier comment](https://github.com/golang/go/issues/71203#issuecomment-2591584886).

Subjectively, I think the generated summary was easier to consume than the huge list of links I ended up with. However, I must admit that I didn't actually try following most of the links because I thought I could guess what they were linking to based on what I already had in my head from writing my summary manually.

The "off-by-one" errors in the linking are unfortunate, but I did still enjoy the text summary of topics that had already been raised since GitHub had already hidden most of the discussion and so the issue degenerated into the same comments made over and over again. It seems like that slowed a little after the summary appeared and the discussion switched to some new topics, but the summary is now also hidden by GitHub and so duplicate remarks seem to have started again. It's unfortunate that GitHub doesn't have a concept of "pinned comments" that are prioritized for display regardless of how many other comments there are.

**Reply by rogpeppe** [reply]
FWIW as someone that hasn't read the full discussion but is familiar with other similar previous proposals, I felt that the gabybot summary was useful and saved me a bunch of time. I feel that I know where the discussion has gone and know enough to decide whether I would like to contribute or dive deeper.

**Comment by seankhliao** [comment]
I think it should ignore any issues closed/locked as spam
https://github.com/golang/go/issues/72051#issuecomment-2692113823

**Comment by firelizzard18** [comment]
I would find a search feature very useful. GitHub's issue search is not very useful, and it soulds like Gaby's database includes a lot more than just issues. I can't count the number of times there's some obscure feature or issue report or something that I remember reading, but barely, and I have a really hard time finding it again. Even if it was as simple as typing in a blurb and having Gaby treat it as a new issue to check for duplicates would be valuable.

**Comment by randall77** [comment]
Gaby can sometimes be late and redundant to a conversation. For example, on #73253 @thepudds posted saying "I sent CL 663795 for discussion", and then Gaby commented a day or so later mentioning that CL as a "related code change".
Probably not worth the comment if a human already posted that observation.

**Comment by toqueteos** [comment]
Regarding the `CL NNNN` _"magic text replacement"_, GitHub does already provide ways to do that without a bot.

They call it [Autolink references](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls#custom-autolinks-to-external-resources)

It's accessible to repo admins from https://github.com/ORG/REPO/settings/key_links

**Reply by jakebailey** [reply]
Not for `CL NNNN`, only `CLNNNN` or `CL-NNNN` (no spaces allowed).

**Comment by dolmen** [comment]
GabyHelp has added the `ToolProposal` label to #74786, but this isn't relevant. I have not even a mention of "tool" in the title or description.

Could it be that it added that label because the issues I have opened recently _are_ related to the ToolProposal? If yes, this is incorrect behavior.

**Reply by dolmen** [reply]
Cc: @jba

**Comment by AGWA** [comment]
It would nice if gabyhelp comments were automatically deleted if they receive a sufficient number of negative reactions (maybe `numPositive - numNegative <= -2`).  That way unhelpful comments wouldn't clutter up the issue.

(I got this idea from Reddit bots whose comments self-destruct if they are downvoted.)

**Comment by Grax23** [comment]
These hackathons are getting old
